examples/run_smc_on_freesolv.py

```diff
@@ -69,16 +150,53 @@ def run_on_freesolv_mol(mol):
     print(f"predicted: {dG:.3f} kcal/mol")
     print(f"experimental: {props['dG']:.3f} kcal/mol")
 
+    initial_samples_refined, final_samples_refined = get_endstate_samples_from_smc_result(
+        smc_result, propagate, lambdas
+    )
+    smc_result["initial_samples_refined"] = initial_samples_refined
+    smc_result["final_samples_refined"] = final_samples_refined
+
     # save summary
-    uid = f"mol={name}_time={datetime.now().isoformat(timespec='seconds')}"
-    save_smc_result(uid, smc_result, save_full_trajectories=cmd_args.debug_mode)
+    save_smc_result(result_path, mol, smc_result, cmd_args, save_full_trajectories=cmd_args.debug_mode)
 
 
-if __name__ == "__main__":
+def run_on_mols(mols: List[Chem.rdchem.Mol], cmd_args: argparse.Namespace):
+    for mol in mols:
+        run_on_freesolv_mol(mol, cmd_args)
+
+
+def main():
     cmd_args = parse_options()
+    mols = fetch_freesolv(n_mols=cmd_args.n_mols, filter_mols=cmd_args.filter_mols)
 
-    mols = fetch_freesolv()
-    select_mols = mols[: cmd_args.n_mols]
+    # Create result folder
+    result_path = Path(cmd_args.result_path)
+    result_path.mkdir(exist_ok=True, parents=True)
 
-    for mol in select_mols:
-        run_on_freesolv_mol(mol)
+    # Set up client
+    num_gpus = cmd_args.n_gpus or 1
+    cmd_args.n_cpus = cmd_args.n_cpus or os.cpu_count()
+
+    if num_gpus > 1:
+        client = CUDAPoolClient(max_workers=num_gpus)
+        cmd_args.n_cpus = cmd_args.n_cpus // num_gpus
+    else:
+        client = SerialClient()
+    client.verify()
+    print(f"using {num_gpus} gpus with {cmd_args.n_cpus} cpus per gpu")
+
+    # Batch mols
+    batch_mols = defaultdict(list)
+    for i in range(len(mols)):
+        batch_mols[i % num_gpus].append(mols[i])
+
+    futures = []
+    for mol_subset in batch_mols.values():
+        futures.append(client.submit(run_on_mols, mol_subset, cmd_args))
+
+    # Wait for jobs to complete
+    _ = [fut.result() for fut in futures]
+
```

Comment:
Some of this logic looks generic enough to consider moving into the parallel module at some point